Fixes issue #2208

vigneashs opened the pull request:
This PR adds additional validation to authentication path computation for secondary requests.

blueww commented:
@vigneashs 

Thanks for raising the fix PR!
Would you please also add:
1. Change log: https://github.com/Azure/Azurite/blob/main/ChangeLog.md
2. Test case: you can refer this case for how to add test case with product style Url: https://github.com/Azure/Azurite/blob/2fc3a7b8533a84551c7a5acc2ec95581785b5043/tests/blob/specialnaming.test.ts#L448

blueww commented:
For the test case, I have drafted a test case in https://github.com/Azure/Azurite/pull/2216/files, for your refrence.

blueww commented on ChangeLog.md:
```diff
@@ -12,15 +12,18 @@ Blob:
 - Fixed startCopyFromURL, copyFromURL API to return 400 (InvalidHeaderValue) when copy source has invalid format. (issue #1954)
 - Fixed CommitBlockList API to return 400 (InvalidXmlDocument) when the request is sent with JSON body. (issue #1955)
 - Added "x-ms-is-hns-enabled" header in x-ms-is-hns-enabled API responds (issue #1810)
+- Fixed authentication error in production style URL (issue #2208)
```
Should mention secondary in the changelog. Like:
- Fixed authentication error in production style URL for secondary location (issue #2208)

Same for T/Q change log.

blueww commented:
@vinagesh 

The following new case will fail, would you please help to look?
```
  1) table name validation tests
       Should work with production style URL when devstoreaccount1-secondary.table.localhost is resolvable:
     AssertionError [ERR_ASSERTION]: Failed
      at /home/vsts/work/1/s/tests/table/apis/table.validation.rest.test.ts:457:18
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at async Context.<anonymous> (tests/table/apis/table.validation.rest.test.ts:441:5)
```

vigneashs commented:
> @vinagesh
> 
> The following new case will fail, would you please help to look?
> 
> ```
>   1) table name validation tests
>        Should work with production style URL when devstoreaccount1-secondary.table.localhost is resolvable:
>      AssertionError [ERR_ASSERTION]: Failed
>       at /home/vsts/work/1/s/tests/table/apis/table.validation.rest.test.ts:457:18
>       at runMicrotasks (<anonymous>)
>       at processTicksAndRejections (internal/process/task_queues.js:95:5)
>       at async Context.<anonymous> (tests/table/apis/table.validation.rest.test.ts:441:5)
> ```

I will fix it. Thanks

vigneashs commented:
@blueww, I have fixed the error in Table unit tests. 

The createStringToSignForSharedKeyLite routine had a hardcoded storage account name in authentication path which is valid only for non-production style URL. I have fixed that method to use both production style URL and non-production style URL.

blueww commented:
/azp run

blueww commented:
@XiaoningLiu , @EmmaZhu  
This PR looks good to me.
Would you please help to review this PR and see if you have any comments?
I plan to merge it in 1-2 days.
